chaosazure/vmss/fetcher.py

```diff
@@ -26,62 +41,54 @@ def choose_vmss_instance_at_random(vmss_choice, configuration, secrets):
-    vmss_instances = fetch_vmss_instances(vmss_choice, configuration, secrets)
-    if not vmss_instances:
-        logger.debug("No virtual machine scale set instances found")
-        raise FailedActivity("No virtual machine scale set instances found")
-    else:
-        logger.debug(
-            "Found virtual machine scale set instances: {}".format(
-                [x['name'] for x in vmss_instances]))
-
-    choice_vmss_instance = None
-    for vmss in vmss_instances:
-        if vmss_matches_criteria(vmss, instance_criteria):
-            choice_vmss_instance = vmss
-            break
+                         secrets: Secrets = None) -> List[Dict[str, Any]]:
+    result = []
+    instances = fetch_vmss_instances(vmss_choice, configuration, secrets)
 
-    if not choice_vmss_instance:
-        logger.debug("No virtual machine scale set instance found for "
-                     "virtual machine scale set %s and criteria %s"
-                     % (vmss, instance_criteria))
-        raise FailedActivity("No virtual machine scale set instances found for"
-                             " criteria")
+    for instance in instances:
+        if is_criteria_matched(instance, instance_criteria):
+            result.append(instance)
 
-    logger.warning("Attempting to stop instance with name %s"
-                   % choice_vmss_instance['name'])
+    if len(result) == 0:
+        msg = "No VMSS instance found for criteria '{}'".format(
+            instance_criteria)
+        logger.debug(msg)
+        raise FailedActivity(msg)
 
-    return choice_vmss_instance
+    return result
 
 
-def vmss_matches_criteria(vmss: dict,
-                          instance_criteria:
-                          Iterable[Mapping[str, any]] = None):
+def is_criteria_matched(instance: dict,
+                        instance_criteria:
+                        Iterable[Mapping[str, any]] = None):
     for criteria in instance_criteria:
-        logger.debug("Checking criteria %s" % criteria)
-        found_mismatch = False
+        found = False
+        mismatch = False
         for key, value in criteria.items():
-            if vmss[key] != value:
-                found_mismatch = True
-                break
-        if not found_mismatch:
-            logger.debug("Matching criteria %s" % criteria)
+            if instance[key] == value:
+                found = True
+            else:
+                mismatch = True
+
+        if found and not mismatch:
             return True
 
     return False
 
 
-def choose_vmss_at_random(filter, configuration, secrets):
+def fetch_vmss(filter, configuration, secrets) -> []:
     vmss = fetch_resources(filter, RES_TYPE_VMSS, secrets, configuration)
+
     if not vmss:
-        logger.warning("No virtual machine scale sets found")
-        raise FailedActivity("No virtual machine scale sets found")
+        msg = "No VMSS found"
+        logger.warning(msg)
+        raise FailedActivity(msg)
```

Comment:
As well here I have a feeling that we will log the same message twice

Reply:
Solved